### PR TITLE
Create migrations and models for users, incidents, and attachments

### PR DESCRIPTION
**Pull request by jhankin**

### CODE REVIEW

File: app/models/incident.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class Incident < ApplicationRecord
+  belongs_to :user
```
Comment (seansu4you87) [inline app/models/incident.rb]: add inverse `has_many :attachments`

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+class User < ApplicationRecord
+end
```
Comment (seansu4you87) [inline app/models/user.rb]: add inverse `has_many :incidents`

File: db/migrate/20170426040953_create_users.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class CreateUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :users do |t|
+      t.integer :facebook_id
```
Comment (seansu4you87) [inline db/migrate/20170426040953_create_users.rb]: `unique: true, null: false`

File: db/migrate/20170426040953_create_users.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+      t.string :gender
+      t.string :race
+      t.timestamp :birthdate
```
Comment (seansu4you87) [inline db/migrate/20170426040953_create_users.rb]: `t.datetime :born_at` to follow rails database modeling convention

File: db/migrate/20170426041307_create_incidents.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+class CreateIncidents < ActiveRecord::Migration[5.0]
+  def change
+    create_table :incidents do |t|
+      t.references :user, foreign_key: true
```
Comment (seansu4you87) [inline db/migrate/20170426041307_create_incidents.rb]: `null: false`

File: db/migrate/20170426041307_create_incidents.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+      t.float :latitude
+      t.float :longitude
+      t.integer :rating
+      t.string :incident_type
```
Comment (seansu4you87) [inline db/migrate/20170426041307_create_incidents.rb]: nit: `type` is prolly clear enough

File: db/migrate/20170426041307_create_incidents.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+      t.string :tags, array: true
```
Comment (seansu4you87) [inline db/migrate/20170426041307_create_incidents.rb]: what are these tags?  do we need to denormalize these now?

File: db/migrate/20170426041307_create_incidents.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+      t.boolean :finished
+      t.boolean :is_fake
```
Comment (seansu4you87) [inline db/migrate/20170426041307_create_incidents.rb]: nit: `fake` instead, to fall in line with `finished`

File: db/migrate/20170426041307_create_incidents.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+      t.json :metadata
```
Comment (seansu4you87) [inline db/migrate/20170426041307_create_incidents.rb]: `null: false`, we should always be saving some kind of metadata

File: db/migrate/20170426042354_create_attachments.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+class CreateAttachments < ActiveRecord::Migration[5.0]
+  def change
+    create_table :attachments do |t|
+      t.references :incident, foreign_key: true
```
Comment (seansu4you87) [inline db/migrate/20170426042354_create_attachments.rb]: `null: false`

File: db/migrate/20170426042354_create_attachments.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+      t.string :url
```
Comment (seansu4you87) [inline db/migrate/20170426042354_create_attachments.rb]: `null: false`

Also, `unique: true`?  Not totally sure on this one, but it seems about right

File: test/models/attachment_test.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (seansu4you87) [inline test/models/attachment_test.rb]: Let's delete all of these test files, since we're using rspec

File: test/models/incident_test.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (seansu4you87) [inline test/models/incident_test.rb]: kill

File: test/models/user_test.rb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (seansu4you87) [inline test/models/user_test.rb]: kill

File: test/fixtures/users.yml [hunk 15]
```diff
@@ -0,0 +1,13 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (seansu4you87) [inline test/fixtures/users.yml]: Are we using these fixtures?  If not explicitly, let's kill these.  We're likely using factories instead

File: db/migrate/20170426040953_create_users.rb (hunk 3 above)
Comment (jhankin) [inline db/migrate/20170426040953_create_users.rb]: Also changing this to `t.string`, found `t.integer` was a problem when i tested using actual facebook ids.

@seansu4you87 will it be a problem to remove the `null: false` later if we add a second chatbot client that doesn't use facebook ids?

File: db/migrate/20170426041307_create_incidents.rb (hunk 6 above)
Comment (jhankin) [inline db/migrate/20170426041307_create_incidents.rb]: in reality, they're both too vague 😂 Maybe `initial_contact`?

File: db/migrate/20170426041307_create_incidents.rb (hunk 7 above)
Comment (jhankin) [inline db/migrate/20170426041307_create_incidents.rb]: We don't have to yet, I was picturing we do reactions, feelings, etc. all as tags for ease of use, but you're right, we're not planning to search by them immediately.

File: db/migrate/20170426042354_create_attachments.rb (hunk 11 above)
Comment (jhankin) [inline db/migrate/20170426042354_create_attachments.rb]: yeah i dig it

File: db/migrate/20170426041307_create_incidents.rb (hunk 9 above)
Comment (jhankin) [inline db/migrate/20170426041307_create_incidents.rb]: OK, gonna add `default: {}` as well since we will probably create them with only a `facebook_id`

**Comment by jhankin** [comment]
@seansu4you87 updated, PTAL

File: db/migrate/20170426040953_create_users.rb [hunk 16]
```diff
@@ -0,0 +1,12 @@
+class CreateUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :users do |t|
+      t.string :facebook_id, unique: true, null: false
+      t.string :gender
+      t.string :race
```
Comment (seansu4you87) [inline db/migrate/20170426040953_create_users.rb]: For gender, race and birthday, what happens when between incidents, these change?  (We don't necessarily have to solve this problem now)

File: db/schema.rb [hunk 17]
```diff
@@ -10,9 +10,41 @@
...
+    t.boolean  "finished"
+    t.boolean  "fake"
+    t.json     "metadata",      default: {}, null: false
+    t.datetime "created_at",                 null: false
+    t.datetime "updated_at",                 null: false
+    t.index ["user_id"], name: "index_incidents_on_user_id", using: :btree
+  end
+
+  create_table "users", force: :cascade do |t|
+    t.string   "facebook_id", null: false
```
Comment (seansu4you87) [inline db/schema.rb]: Was the scheme updated? How come there's no indication of unique indices?

File: db/migrate/20170426041307_create_incidents.rb [hunk 18]
```diff
@@ -0,0 +1,16 @@
+class CreateIncidents < ActiveRecord::Migration[5.0]
+  def change
+    create_table :incidents do |t|
+      t.references :user, foreign_key: true, null: false
+      t.float :latitude
+      t.float :longitude
+      t.integer :rating
+      t.string :incident_type
```
Comment (seansu4you87) [inline db/migrate/20170426041307_create_incidents.rb]: nit: naming here

Comment (seansu4you87) [inline db/migrate/20170426041307_create_incidents.rb]: u mentioned `initial_contact`.  That also confused me lol.  What exactly can the options be here?

**Comment by seansu4you87** [comment]
Also note for later, we'll need to add indices on `incidents` for things that are being queried by the dashboard.  This can happen later though, when the Dashboard has more usage

File: db/migrate/20170426041307_create_incidents.rb (hunk 18 above)
Comment (jhankin) [inline db/migrate/20170426041307_create_incidents.rb]: Yeah, so, what we're calling "incident type" today is one of:

* got pulled over
* called the police
* got stopped on foot
* witnessed police
* other

Which is why @Jlawlzz at one point suggested calling it "initial contact", which I still think is a good idea.  There's also a separate question which we currently save as a "tag" which is like "Citation", "Police Shooting," "Conversation," "Arrest", etc., which kinda sounds more like an "incident type" to me.  Thoughts?

File: db/migrate/20170426040953_create_users.rb (hunk 16 above)
Comment (jhankin) [inline db/migrate/20170426040953_create_users.rb]: We page?  I dunno, it shouldn't change, but people should be able to correct their mistakes (or change their gender if they transition or something).

File: db/schema.rb (hunk 17 above)
Comment (jhankin) [inline db/schema.rb]: The syntax I used didn't work. Take another look? should be good now

**Comment by seansu4you87** [comment]
lgtm

**Pull request merged**